File: c/problem_34.c
#include "problem_34.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

#define MAXN 64

static int dp34[MAXN][MAXN];
static const char *gs34;

static void build34(int i, int j, char *out, int *pos) {
    if (i > j) return;
    if (i == j) { out[(*pos)++] = gs34[i]; return; }
    if (gs34[i] == gs34[j]) {
        out[(*pos)++] = gs34[i];
        build34(i+1, j-1, out, pos);
        out[(*pos)++] = gs34[j];
        return;
    }
    char left[MAXN*2], right[MAXN*2];
    int lp=0, rp=0;
    left[lp++] = gs34[j];  build34(i, j-1, left, &lp);  left[lp++] = gs34[j];  left[lp]='\0';
    right[rp++] = gs34[i]; build34(i+1, j, right, &rp); right[rp++] = gs34[i]; right[rp]='\0';
    const char *chosen;
    if      (dp34[i+1][j] < dp34[i][j-1]) chosen = right;
    else if (dp34[i][j-1] < dp34[i+1][j]) chosen = left;
    else                                    chosen = strcmp(left,right) <= 0 ? left : right;
    int len = (int)strlen(chosen);
    memcpy(out + *pos, chosen, len);
    *pos += len;
}

int minPalindrome(const char *s, char *out) {
    int n = (int)strlen(s);
    memset(dp34, 0, sizeof(dp34));
    for (int len = 2; len <= n; len++)
        for (int i = 0; i <= n - len; i++) {
            int j = i + len - 1;
            if (s[i] == s[j]) dp34[i][j] = (len==2) ? 0 : dp34[i+1][j-1];
            else {
                int a = dp34[i+1][j], b = dp34[i][j-1];
                dp34[i][j] = 1 + (a < b ? a : b);
            }
        }
    gs34 = s;
    int pos = 0;
    build34(0, n-1, out, &pos);
    out[pos] = '\0';
    return pos;
}

void problem34(void) {
    printf("Problem 34:\n");
    char buf[256];
    minPalindrome("race", buf);   assert(strcmp(buf, "ecarace")  == 0);
    minPalindrome("google", buf); assert(strcmp(buf, "elgoogle") == 0);
    minPalindrome("a", buf);      assert(strcmp(buf, "a")        == 0);
    minPalindrome("ab", buf);     assert(strcmp(buf, "aba")      == 0);
    printf("All Tests Passed\n");
}

File: rust/problem_34.rs
pub fn min_palindrome(s: &str) -> String {
    let s: Vec<char> = s.chars().collect();
    let n = s.len();
    let mut dp = vec![vec![0usize; n]; n];
    for len in 2..=n {
        for i in 0..=(n - len) {
            let j = i + len - 1;
            dp[i][j] = if s[i] == s[j] {
                if len == 2 { 0 } else { dp[i+1][j-1] }
            } else {
                1 + dp[i+1][j].min(dp[i][j-1])
            };
        }
    }
    fn build(s: &[char], dp: &Vec<Vec<usize>>, i: usize, j: usize) -> String {
        if i > j { return String::new(); }
        if i == j { return s[i].to_string(); }
        if s[i] == s[j] {
            return format!("{}{}{}", s[i], build(s, dp, i+1, j-1), s[j]);
        }
        let add_left  = format!("{}{}{}", s[j], build(s, dp, i, j-1), s[j]);
        let add_right = format!("{}{}{}", s[i], build(s, dp, i+1, j), s[i]);
        if dp[i+1][j] < dp[i][j-1] { return add_right; }
        if dp[i][j-1] < dp[i+1][j] { return add_left; }
        add_left.min(add_right)
    }
    build(&s, &dp, 0, n - 1)
}

pub fn problem34() {
    println!("Problem 34:");
    assert_eq!(min_palindrome("race"),   "ecarace");
    assert_eq!(min_palindrome("google"), "elgoogle");
    assert_eq!(min_palindrome("a"),      "a");
    assert_eq!(min_palindrome("ab"),     "aba");
    println!("All Tests Passed");
}
